indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/PendingTaskBasedResourceManagementStrategy.java

```diff
@@ -0,0 +1,360 @@
+              laziestWorkerIps.size(),
+              Joiner.on(", ").join(laziestWorkerIps)
+          );
+
+          final AutoScalingData terminated = workerConfig.getAutoScaler().terminate(laziestWorkerIps);
+          if (terminated != null) {
+            currentlyTerminating.addAll(terminated.getNodeIds());
+            lastTerminateTime = new DateTime();
+            scalingStats.addTerminateEvent(terminated);
+            didTerminate = true;
+          }
+        }
+      } else {
+        Duration durSinceLastTerminate = new Duration(lastTerminateTime, new DateTime());
+
+        log.info("%s terminating. Current wait time: %s", currentlyTerminating, durSinceLastTerminate);
+
+        if (durSinceLastTerminate.isLongerThan(config.getMaxScalingDuration().toStandardDuration())) {
+          log.makeAlert("Worker node termination taking too long!")
+             .addData("millisSinceLastTerminate", durSinceLastTerminate.getMillis())
+             .addData("terminatingCount", currentlyTerminating.size())
+             .emit();
+
+          currentlyTerminating.clear();
+        }
+      }
+
+      return didTerminate;
+    }
+  }
+
+  private int getScaleDownTarget(Collection<ZkWorker> zkWorkers, WorkerBehaviorConfig workerConfig)
+  {
+    final Predicate<ZkWorker> isValidWorker = ResourceManagementUtil.createValidWorkerPredicate(config);
+    final int currValidWorkers = Collections2.filter(zkWorkers, isValidWorker).size();
+    final int invalidWorkers = zkWorkers.size() - currValidWorkers;
+    final int minWorkers = workerConfig.getAutoScaler().getMinNumWorkers();
+
+    // Scale Down target
+    // All invalid workers + any lazy workers above minCapacity
+    return invalidWorkers + Math.max(
+        0,
+        Math.min(
+            config.getMaxScalingStep(),
+            currValidWorkers - minWorkers
+        )
+    );
+  }
+
+  @Override
+  public ScalingStats getStats()
+  {
+    return scalingStats;
+  }
+
+  private static int getAverageWorkerCapacity(final Collection<ZkWorker> zkWorkers)
+  {
+    int size = zkWorkers.size();
+    if (size == 0) {
+      // No existing workers assume capacity per worker as 1
+      return 1;
+    } else {
+      // Assume all workers have same capacity
+      return zkWorkers.iterator().next().getWorker().getCapacity();
+    }
+  }
+
+  private static ImmutableZkWorker workerWithTask(ImmutableZkWorker immutableZkWorker, Task task)
+  {
+    return new ImmutableZkWorker(
+        immutableZkWorker.getWorker(), immutableZkWorker.getCurrCapacityUsed() + 1, Sets.union(
+        immutableZkWorker.getAvailabilityGroups(), Sets.newHashSet(
+            task.getTaskResource()
+                .getAvailabilityGroup()
+        )
+    )
+    );
+  }
+
+  private static ImmutableZkWorker createDummyWorker(String host, int capacity, String version)
+  {
+    return new ImmutableZkWorker(new Worker(host, "-1", capacity, version), 0, Sets.<String>newHashSet());
```

Comment:
Note: https://groups.google.com/d/msg/druid-development/TGTLjG2H_Mk/VgZrp9drBQAJ

Should this be "-2"?